### PR TITLE
8282469: Allow considered use of C++ thread_local in Hotspot

### PR DESCRIPTION
**Pull request by dholmes-ora**
This patch provides a means for using C++ `thread_local` when it is essential - see JBS for more details.

There are three parts:

1. Add the new #define for `thread_local`
2. Remove `operator_new.cpp` as use of C++ `thread_local` with a non-trival cleanup actions requires use of global operators new/delete. These are still excluded for hotspot use via a link-time check.
3. Remove the prohibition on using `thread_local` from the hotspot style guide

Due to the way hotspot style guide changes must be done, part 3 is being done under a sub-task in PR https://github.com/openjdk/jdk/pull/7720 and the two PR's will integrate at the same time.

Testing:
  - manual testing of the Panama usecase as referenced in the JBS issue
  - Tiers 1-3

Thanks,
David

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282469](https://bugs.openjdk.java.net/browse/JDK-8282469): Allow considered use of C++ thread_local in Hotspot

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Daniel D. Daugherty](https://openjdk.java.net/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7719/head:pull/7719` \
`$ git checkout pull/7719`

Update a local copy of the PR: \
`$ git checkout pull/7719` \
`$ git pull https://git.openjdk.java.net/jdk pull/7719/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7719`

View PR using the GUI difftool: \
`$ git pr show -t 7719`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7719.diff">https://git.openjdk.java.net/jdk/pull/7719.diff</a>

</details>

### CODE REVIEW

**Review by kimbarrett — APPROVED** [review]
Looks good.

File: src/hotspot/share/memory/operator_new.cpp [hunk 1]
```diff
@@ -1,100 +0,0 @@
...
-#include <new>
-
-//--------------------------------------------------------------------------------------
-// Non-product code
-
-#ifndef PRODUCT
-// The global operator new should never be called since it will usually indicate
-// a memory leak.  Use CHeapObj as the base class of such objects to make it explicit
-// that they're allocated on the C heap.
-// Commented out in product version to avoid conflicts with third-party C++ native code.
```
Comment (rose00) [inline src/hotspot/share/memory/operator_new.cpp:37]: There's a little bit of policy information here that is being deleted.  It overlaps with the section `### Memory Allocation` in `hotspot-style.md`, but includes this information which might not be stated elsewhere:

> Typically, uses of the C++ global operator new are inadvertent and therefore often associated with memory leaks.

(This is my rephrasing, perhaps appropriate to the style guide.)

Or is a point like this made in the config file which prevents direct linkage?  (I don't know where that file is.)

Comment (dholmes-ora) [reply src/hotspot/share/memory/operator_new.cpp:37]: The link-time check is expressed in open/make/hotspot/lib/CompileJvm.gmk but simply states:

# Hotspot disallows the use of global operators 'new' and 'delete'. This build
# time check helps enforce this requirement. ...

The prohibition on using global operator new should definitely be explicitly documented somewhere, so the style guide seems a suitable place.

**Comment by dholmes-ora** [comment]
Thanks for the review @kimbarrett !

File: src/hotspot/share/memory/operator_new.cpp (hunk 1 above)
Comment (dcubed-ojdk) [reply src/hotspot/share/memory/operator_new.cpp:37]: I agree that the prohibition on using global operator new should definitely be explicitly documented somewhere and I'm good with adding that warning to the HotSpot style guide.

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.

I had to reread these two sentences a couple of times:

2a) Remove operator_new.cpp as use of C++ thread_local with a non-trival cleanup actions requires use of global operators new/delete.
2b) These are still excluded for hotspot use via a link-time check.

I think I've convinced myself that the link-time check won't run afoul of the
uses that are planned for Panama. I did that by rephrasing the above into:

> Panama's use of thread_local will have non-trival cleanup actions and those
> will not be complained about by the link-time check.

Please let me know if I have this correct.

**Comment by dholmes-ora** [comment]
> Please let me know if I have this correct.

@dcubed-ojdk  you statement is not wrong but it an odd way of rephrasing things. The current change  has no affect on the link-time check, so if a panama use would run afoul of that check before, then it will still run afoul of it. But I don't expect it to run afoul in either case.

The problem I'm fixing with this part of the change is that the C++ library will currently link to our (booby-trapped) definitions of global operators new and delete, where it needs the real ones, so I deleted our definitions. We still don't want hotspot code to use these global external definitions though and that is where the pre-existing link-time check comes in - we check all libjvm component object files for symbols referring to the disallowed global methods.

Hope that clarifies.

**Comment by dcubed-ojdk** [comment]
Thanks for the clarification. It does help.

**Comment by dholmes-ora** [comment]
Style guide update has been approved and integrated.

/integrate

**Pull request closed without merging**